### PR TITLE
Segmenting-DMD

### PR DESCRIPTION
**Pull request by Jimmy-INL**
Making time segmentation work for DMD.

--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Rime segmentation only worked for ARMA, this is an attempt to generalize it to other time-dependent surrogate models.

##### What are the significant changes in functionality due to this change request?
User build a sequence of DMD surrogates based on a user-selection of number of divisions or length(s) of divisions.
##### TODO:
1. allow user to input different DMD specs for each segment, such as different ranks, modes, and/or optimized modes selection.
2. It will be great of the algorithm is extended to be smart enough to determine the time windows to fit the best number of segments and size of segments which can be learned from temporal bahavious as a function of the sampled input parameters.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/SupervisedLearning/ROMCollection.py
Comment (JiaZhou-PU) [inline framework/SupervisedLearning/ROMCollection.py:532]: I am not sure about the data type in here, it looks like a list before but array now? I am fine with array though.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (JiaZhou-PU) [inline framework/SupervisedLearning/SupervisedLearning.py:85]: Why removing this check?

**Review by JiaZhou-PU — COMMENTED** [review]
Looks great

File: framework/SupervisedLearning/ROMCollection.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/ROMCollection.py:532]: I think the `TODO` is resolved by this change, probably, so we can remove the `TODO` and leave the rest of the comment.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/SupervisedLearning.py:85]: I agree; how did this pass for DMD before this PR? Was it only ever trained on single histories?

File: framework/SupervisedLearning/DynamicModeDecomposition.py
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/DynamicModeDecomposition.py:94]: Please add a comment about why this "if" was added. It's generally not okay to deepcopy an untrained SVL engine ... except we do it in the ROMCollection, so that's the exception. That comment might save someone a lot of searching someday.

File: framework/SupervisedLearning/ROMCollection.py
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/ROMCollection.py]: Is there every a check in `Clustered` that disables the ability to specify different lengths? I believe it calls this same `subdivideDomain` method, but without that check this would get messy fast without a good error.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/SupervisedLearning.py]: Is this right? It looks like the argument to `np.zeros` might look like
```python
  np.zeros( [(2,3), 4] )
```
(square brackets for clarity, it's actually all tuples). Is zeros okay with mixed tuples? Either way, this could probably be clearer.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/SupervisedLearning.py:211]: This seems odd to me. What is the indexing or shape of valueToUse versus featureValues? At minimum there should be more comments here to explain this check.

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
A few things to spruce up. I realize most of the problems were inherited from older implementation, but I want to leave this so the next person doesn't have to struggle to make sense of it, at least not as much as before.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (Jimmy-INL) [reply framework/SupervisedLearning/SupervisedLearning.py:85]: It did not. It always gave me an error here. the array is always bigger than 1-d

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (Jimmy-INL) [reply framework/SupervisedLearning/SupervisedLearning.py:211]: This change changes nothing, but I prefer to use shape with ndarrays than len. In Arma, the first dimension is always 1, but in other t-dependent surroagates this will be the number of samples. I will probably add a comment though.

File: framework/SupervisedLearning/DynamicModeDecomposition.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/DynamicModeDecomposition.py:94]: 👍

File: framework/SupervisedLearning/ROMCollection.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/ROMCollection.py:532]: 👍

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/SupervisedLearning.py:85]: 👍 
Note we have to revisit this if/when we start training ROMs on ND input data; for now this is not a thing.

File: framework/SupervisedLearning/SupervisedLearning.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/SupervisedLearning.py:211]: 👍

**Review by PaulTalbot-INL — APPROVED** [review]
Looks good to me. Since this all has to be reviewed before merging into devel, this stage of contribution is fine for now.

**Comment by PaulTalbot-INL** [comment]
I'll hold off merging this for now until a test case can be added.

**Comment by PaulTalbot-INL** [comment]
@Jimmy-INL Does this PR still have a purpose? note that it points to the ARMA multiyear rework branch, which has been merged to devel of main raven

**Comment by Jimmy-INL** [comment]
I think I need to rebase and then generate anew PR on devel

**Pull request closed without merging**
